### PR TITLE
Isolate cluster operator in dedicated namespace, and other resources …

### PR DESCRIPTION
**Pull request by see-quick**
…have its own namespace

### Type of change

- Enhancement / new feature
- Refactoring

### Description

This PR adds a dedicated namespace for Cluster Operator and other namespaces for resources (e.g., Kafka, KafkaBridge...). Currently, we have a Cluster Operator namespace (i.e., infra-namespace), which shares space with test suite resources. But in an ideal world, we would like to have the following:
1.  dedicated namespace only for Cluster Operator 
2. test suite namespace for test class resources such as Kafka, KafkaBridge, KafkaConnect, and KMM2.
3. and lastly namespace for test cases, which needs to run KMM2 or Kafka instances in some form of isolation and needs its own namespace (i.e., @ParallelNamespaceTest).

And so these three points mentioned above bring this PR. 

Worth mentioning is that for `upgrade and downgrade test cases` I have let stick with one namespace (i.e., cluster operator one). 

### Checklist

- [x] Write tests
- [x] Kraft-regression is working
- [x] Bundle regression
- [x] Bundle-rbac=NAMESPACE acceptance
- [x] HELM acceptance
- [x] OLM acceptance
- [x] Upgrade
- [x] FeatureGates regression

### CODE REVIEW

**Comment by see-quick** [comment]
@strimzi-ci run tests --profile=acceptance --install-type=olm --ocp-version=4.13 --env=OLM_OPERATOR_NAME=strimzi-kafka-operator,OLM_OPERATOR_VERSION=6.6.6,OLM_SOURCE_NAME=strimzi-source,OLM_OPERATOR_DEPLOYMENT_NAME=strimzi-kafka-operator

**Comment by see-quick** [comment]
@strimzi-ci run tests --profile=acceptance --install-type=olm --ocp-version=4.13 --env=OLM_OPERATOR_NAME=strimzi-kafka-operator;OLM_OPERATOR_VERSION=6.6.6;OLM_SOURCE_NAME=strimzi-source;OLM_OPERATOR_DEPLOYMENT_NAME=strimzi-kafka-operator

**Review by henryZrncik — APPROVED** [review]

File: systemtest/src/main/java/io/strimzi/systemtest/Constants.java [hunk 1]
```diff
@@ -409,8 +409,8 @@ public interface Constants {
     String LOAD_BALANCER_CLEANUP = "service.kubernetes.io/load-balancer-cleanup";
 
     // main namespace for Cluster Operator deployment
-    String INFRA_NAMESPACE = "infra-namespace";
-    String METRICS_SECOND_NAMESPACE = "second-metrics-cluster-test";
+    String CO_NAMESPACE = "co-namespace";
+    String TEST_SUITE_NAMESPACE = "test-suite-namespace";
```
[the remark this answers is not in the document]
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/Constants.java:413]: > In case all tests are @ParallelTest, or @IsolatedTest namespace test-suite-namespace is used ?

Yes.

> And apart from that if all tests are @ParallelNamespaceTest we should use co-namespace for cluster operator and something like for respective namespaces ?

You do not need to do it because it's already implemented [1]. Thus, if you use test storage in your test (now convention) there should be implicit `Constants.TEST_SUITE_NAMESPACE`. 

> More or less is this a rule of thumb as you are mentioning briefly that you are keeping co-namespace in some cases?

I keep `co-namespace` for upgrade and downgrade test cases and maybe some edge cases but majority of test cases should follow the same pattern from `TestStorage` (i.e., using `Constants.TEST_SUITE_NAMESPACE`)




[1] - https://github.com/strimzi/strimzi-kafka-operator/pull/8881/files#diff-70e45c4283fca95208daaace1fc0b12726b2d83759b56dd1d6b106a30e9eda71R55

**Review by jankalinic — APPROVED** [review]
Seems good to me. Didn't see anything that would catch my attention or be wrong.

**Comment by see-quick** [comment]
@strimzi-ci run tests --profile=acceptance --install-type=olm --ocp-version=4.13 --env=OLM_OPERATOR_NAME=strimzi-kafka-operator;OLM_OPERATOR_VERSION=6.6.6;OLM_SOURCE_NAME=strimzi-source;OLM_OPERATOR_DEPLOYMENT_NAME=strimzi-kafka-operator

**Comment by see-quick** [comment]
@strimzi-ci run tests --profile=acceptance --install-type=olm --ocp-version=4.13 --env=OLM_OPERATOR_NAME=strimzi-kafka-operator;OLM_OPERATOR_VERSION=6.6.6;OLM_SOURCE_NAME=strimzi-source;OLM_OPERATOR_DEPLOYMENT_NAME=strimzi-kafka-operator

File: systemtest/src/test/java/io/strimzi/systemtest/specific/HelmChartST.java [hunk 2]
```diff
@@ -29,36 +28,32 @@ class HelmChartST extends AbstractST {
 
     @IsolatedTest
     void testStrimziComponentsViaHelmChart(ExtensionContext extensionContext) {
-        String clusterName = mapWithClusterNames.get(extensionContext.getDisplayName());
-        String topicName = mapWithTestTopics.get(extensionContext.getDisplayName());
+        final TestStorage ts = new TestStorage(extensionContext, Constants.TEST_SUITE_NAMESPACE);
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/specific/HelmChartST.java:31]: `testStorage` instead of `ts`? (to have it consistent)

File: systemtest/src/test/java/io/strimzi/systemtest/upgrade/AbstractUpgradeST.java [hunk 3]
```diff
@@ -97,11 +97,11 @@ public class AbstractUpgradeST extends AbstractST {
     protected File kafkaYaml;
 
     protected void makeSnapshots() {
-        coPods = DeploymentUtils.depSnapshot(clusterOperator.getDeploymentNamespace(), ResourceManager.getCoDeploymentName());
-        zkPods = PodUtils.podSnapshot(clusterOperator.getDeploymentNamespace(), zkSelector);
-        kafkaPods = PodUtils.podSnapshot(clusterOperator.getDeploymentNamespace(), kafkaSelector);
-        eoPods = DeploymentUtils.depSnapshot(clusterOperator.getDeploymentNamespace(), KafkaResources.entityOperatorDeploymentName(clusterName));
-        connectPods = PodUtils.podSnapshot(clusterOperator.getDeploymentNamespace(), connectLabelSelector);
+        coPods = DeploymentUtils.depSnapshot(io.strimzi.systemtest.Constants.CO_NAMESPACE, ResourceManager.getCoDeploymentName());
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/upgrade/AbstractUpgradeST.java:100]:
```suggestion
        coPods = DeploymentUtils.depSnapshot(Constants.CO_NAMESPACE, ResourceManager.getCoDeploymentName());
```
(applies for all other places)

**Review by im-konge — COMMENTED** [review]
I left few comments.
During the review, I saw that there is still used `clusterOperator.getDeploymentNamespace()`
Should it be still used? Shouldn't we rather use `Constants.CO_NAMESPACE`, when from now on we should use just this namespace?

File: systemtest/src/main/java/io/strimzi/systemtest/parallel/TestSuiteNamespaceManager.java [hunk 4]
```diff
@@ -36,6 +37,49 @@ public synchronized static TestSuiteNamespaceManager getInstance() {
...
+     */
+    public void createTestSuiteNamespace(ExtensionContext extensionContext) {
+        final String simpleClassName = extensionContext.getRequiredTestClass().getSimpleName();
+
+        if (ExecutionListener.hasSuiteParallelOrIsolatedTest(extensionContext)) {
+            // if RBAC is enabled we don't run tests in parallel mode and with that said we don't create another namespaces
+            if (!Environment.isNamespaceRbacScope()) {
+                KubeClusterResource.getInstance().createNamespace(CollectorElement.createCollectorElement(simpleClassName), Constants.TEST_SUITE_NAMESPACE);
+                NetworkPolicyResource.applyDefaultNetworkPolicySettings(extensionContext, Collections.singletonList(Constants.TEST_SUITE_NAMESPACE));
+                StUtils.copyImagePullSecrets(Constants.TEST_SUITE_NAMESPACE);
```
[the remark this answers is not in the document]
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/parallel/TestSuiteNamespaceManager.java:57]: Noted and I have created a proper issue to have it tracked [1]

[1] - https://github.com/strimzi/strimzi-kafka-operator/issues/8936

File: systemtest/src/main/java/io/strimzi/systemtest/utils/specific/KeycloakUtils.java [hunk 5]
```diff
@@ -18,17 +19,18 @@ private KeycloakUtils() {}
...
      */
-    public static String getToken(String namespaceName, String baseURI, String userName, String password) {
-        String coPodName = kubeClient().getClusterOperatorPodName(namespaceName);
+    public static String getToken(String keycloakNamespace, String baseURI, String userName, String password) {
+        final String testSuiteScraperPodName = ResourceManager.kubeClient().listPodsByPrefixInName(Constants.TEST_SUITE_NAMESPACE, Constants.SCRAPER_NAME).get(0).getMetadata().getName();
+
         return new JsonObject(
-            cmdKubeClient(namespaceName).execInPod(
-                coPodName,
+            cmdKubeClient(keycloakNamespace).execInPod(
```
[the remark this answers is not in the document]
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/utils/specific/KeycloakUtils.java:32]: Yes. The main purpose or this PR was to have Cluster Operator isolated. In the next, I can isolate also Keycloak operator in namespace beside Cluster Operator okay?

File: systemtest/src/test/java/io/strimzi/systemtest/operators/ClusterOperatorRbacST.java [hunk 6]
```diff
@@ -48,27 +49,28 @@ void testCRBDeletionErrorIsIgnoredWhenRackAwarenessIsNotEnabled(ExtensionContext
...
+        cluster.setNamespace(Constants.TEST_SUITE_NAMESPACE);
+
         String coPodName = kubeClient().getClusterOperatorPodName(clusterOperator.getDeploymentNamespace());
         LOGGER.info("Deploying Kafka: {}, which should be deployed even the CRBs are not present", clusterName);
 
         resourceManager.createResourceWithWait(extensionContext, KafkaTemplates.kafkaEphemeral(clusterName, 3).build());
-
         LOGGER.info("CO log should contain some information about ignoring forbidden access to CRB for Kafka");
-        String log = cmdKubeClient().execInCurrentNamespace(Level.DEBUG, "logs", coPodName).out();
+        String log = cmdKubeClient().namespace(clusterOperator.getDeploymentNamespace()).execInCurrentNamespace(Level.DEBUG, "logs", coPodName).out();
```
[the remark this answers is not in the document]
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/operators/ClusterOperatorRbacST.java:66]: Yes, because we are grabbing logs from Cluster Operator. I can change it to `Constants.CO_NAMESPACE`.

File: systemtest/src/test/java/io/strimzi/systemtest/operators/ClusterOperatorRbacST.java [hunk 7]
```diff
@@ -48,27 +49,28 @@ void testCRBDeletionErrorIsIgnoredWhenRackAwarenessIsNotEnabled(ExtensionContext
...
         assertTrue(log.contains("Kafka(" + cmdKubeClient().namespace() + "/" + clusterName + "): Ignoring forbidden access to ClusterRoleBindings resource which does not seem to be required."));
 
         LOGGER.info("Deploying KafkaConnect: {} without rack awareness, the CR should be deployed without error", clusterName);
-        resourceManager.createResourceWithWait(extensionContext, KafkaConnectTemplates.kafkaConnect(clusterName, clusterOperator.getDeploymentNamespace(), 1).build());
+        resourceManager.createResourceWithWait(extensionContext, KafkaConnectTemplates.kafkaConnect(clusterName, Constants.TEST_SUITE_NAMESPACE, 1).build());
 
         LOGGER.info("CO log should contain some information about ignoring forbidden access to CRB for KafkaConnect");
-        log = cmdKubeClient().execInCurrentNamespace(Level.DEBUG, "logs", coPodName).out();
+        log = cmdKubeClient().namespace(clusterOperator.getDeploymentNamespace()).execInCurrentNamespace(Level.DEBUG, "logs", coPodName).out();
```
[the remark this answers is not in the document]
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/operators/ClusterOperatorRbacST.java:73]: Yes, because we are grabbing logs from Cluster Operator. I can change it to Constants.CO_NAMESPACE.

File: systemtest/src/test/java/io/strimzi/systemtest/upgrade/AbstractUpgradeST.java (hunk 3 above)
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/upgrade/AbstractUpgradeST.java:100]: This is a little bit complicated because we use import from `import io.strimzi.api.kafka.model.Constants;`. Thus, our `io.strimzi.systemtest.Constants` uses the same name of the class as from the API module. So I have two ideas:
1. Use our import of constants (i.e., `io.strimzi.systemtest.Constants`) and for others add prefix `io.strimzi.api.kafka.model`.
2. Or we can differentiate our Constants class and rename it to `ConstantsSTs` or something like that.
WDYT @im-konge ?

File: systemtest/src/main/java/io/strimzi/systemtest/utils/specific/KeycloakUtils.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (im-konge) [reply systemtest/src/main/java/io/strimzi/systemtest/utils/specific/KeycloakUtils.java:32]: Yeah okay, makes sense.
Maybe the Keycloak operator can be in same namespace as CO, but we can discuss it later.

File: systemtest/src/test/java/io/strimzi/systemtest/upgrade/AbstractUpgradeST.java (hunk 3 above)
Comment (im-konge) [reply systemtest/src/test/java/io/strimzi/systemtest/upgrade/AbstractUpgradeST.java:100]: `ConstantsSTs` seems to me like another ST class with tests (because of the suffix)
`TestConstants` would make more sense to me

Anyway, this is something for a completely different PR. I didn't see that other import, so let's keep it now as it is.

File: systemtest/src/test/java/io/strimzi/systemtest/operators/ClusterOperatorRbacST.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (im-konge) [reply systemtest/src/test/java/io/strimzi/systemtest/operators/ClusterOperatorRbacST.java:66]: Yeah as I mentioned, I would maybe completely remove the `getDeploymentNamespace()` as it will be always the same. But it can be done in different PR as well.

[the remark this answers is not in the document]
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/operators/ClusterOperatorRbacST.java]: Yeah, make sense :)

**Comment by see-quick** [comment]
> Shouldn't we rather use Constants.CO_NAMESPACE, when from now on we should use just this namespace?

Yes definitely.

File: systemtest/src/main/java/io/strimzi/systemtest/utils/specific/KeycloakUtils.java [hunk 8]
```diff
@@ -180,10 +187,10 @@ public static String updatePolicyOfRealmClient(String namespaceName, String base
      * @return result of creation
      */
     public static String importRealm(String namespaceName, String baseURI, String token, String realmData) {
-        String coPodName = kubeClient(namespaceName).getClusterOperatorPodName();
+        final String testSuiteScraperPodName = ResourceManager.kubeClient().listPodsByPrefixInName(Constants.TEST_SUITE_NAMESPACE, Constants.SCRAPER_NAME).get(0).getMetadata().getName();
```
Comment (Frawless) [inline systemtest/src/main/java/io/strimzi/systemtest/utils/specific/KeycloakUtils.java:190]: wouldn't be better to use some label and list the pod based on it?

File: systemtest/src/test/java/io/strimzi/systemtest/connect/ConnectBuilderST.java [hunk 9]
```diff
@@ -94,6 +94,8 @@ class ConnectBuilderST extends AbstractST {
     private static final String PLUGIN_WITH_OTHER_TYPE_NAME = "plugin-with-other-type";
     private static final String PLUGIN_WITH_MAVEN_NAME = "connector-from-maven";
 
+    private static final String SHARED_KAFKA_CLUSTER = "shared-kafka-connect";
```
Comment (Frawless) [inline systemtest/src/test/java/io/strimzi/systemtest/connect/ConnectBuilderST.java:97]: so is it Kafka cluster or KafkaConnect cluster?

File: systemtest/src/test/java/io/strimzi/systemtest/connect/ConnectBuilderST.java [hunk 10]
```diff
@@ -210,13 +212,13 @@ void testBuildFailsWithWrongChecksumOfArtifact(ExtensionContext extensionContext
 
     @ParallelTest
     void testBuildWithJarTgzAndZip(ExtensionContext extensionContext) {
-        TestStorage testStorage = new TestStorage(extensionContext, clusterOperator.getDeploymentNamespace());
+        TestStorage testStorage = new TestStorage(extensionContext, Constants.TEST_SUITE_NAMESPACE);
```
Comment (Frawless) [inline systemtest/src/test/java/io/strimzi/systemtest/connect/ConnectBuilderST.java:215]: I wonder if the constant shouldn't be part of ResourceManager, it looks strange to me have it in so many places

File: systemtest/src/test/java/io/strimzi/systemtest/kafka/KafkaNodePoolST.java [hunk 11]
```diff
@@ -212,7 +212,7 @@ void setup(ExtensionContext extensionContext) {
 
         clusterOperator = new SetupClusterOperator.SetupClusterOperatorBuilder()
             .withExtensionContext(extensionContext)
-            .withNamespace(INFRA_NAMESPACE)
+            .withNamespace(CO_NAMESPACE)
```
Comment (Frawless) [inline systemtest/src/test/java/io/strimzi/systemtest/kafka/KafkaNodePoolST.java:215]: Everywhere we use `Constants.CO_NAMESPACE`, why not here?

File: systemtest/src/test/java/io/strimzi/systemtest/connect/ConnectBuilderST.java (hunk 10 above)
Comment (Frawless) [reply systemtest/src/test/java/io/strimzi/systemtest/connect/ConnectBuilderST.java:215]: Or maybe directly in Abstract

File: systemtest/src/main/java/io/strimzi/systemtest/utils/specific/KeycloakUtils.java (hunk 8 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/utils/specific/KeycloakUtils.java:190]: I can do it :)

File: systemtest/src/test/java/io/strimzi/systemtest/connect/ConnectBuilderST.java (hunk 9 above)
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/connect/ConnectBuilderST.java:97]: Kafka cluster, good catch thanks

File: systemtest/src/test/java/io/strimzi/systemtest/connect/ConnectBuilderST.java (hunk 10 above)
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/connect/ConnectBuilderST.java:215]: Well, in this case, we do not need to specify such a namespace here. I think this is leftover. By default, we use `Constants.TEST_SUITE_NAMESPACE` in TestStorage if a user does not specify a different namespace.

Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/connect/ConnectBuilderST.java:215]: I have removed it from the whole class.

File: systemtest/src/test/java/io/strimzi/systemtest/kafka/KafkaNodePoolST.java (hunk 11 above)
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/kafka/KafkaNodePoolST.java:215]: True. I will update it :) Thanks

**Review by jankalinic — APPROVED** [review]

**Review by Frawless — APPROVED** [review]

**Review by im-konge — APPROVED** [review]
